python/ray/actor_mesh.py

```diff
@@ -0,0 +1,194 @@
+import copy
+import math
+import os
+import random
+from typing import Any, Callable, Concatenate, Dict, Generic, List, Optional, Tuple, TypeVar, ParamSpec, Union
+from typing import overload
+
+import ray
+from ray.actor import ActorHandle, ActorProxy
+
+T = TypeVar("T", covariant=True)
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+class ActorMeshProxy:
+    def __init__(self, actor_mesh: "ActorMesh") -> None:
+        self.actor_mesh = actor_mesh
+
+    def __getattr__(self, k):
+        return ActorMethodProxy(self.actor_mesh, k)
+    
+class ActorMethod(Generic[P, R]):
+
+    def __init__(self, method):
+        self.method = method
+
+    def all(self, *args: P.args, **kwargs: P.kwargs) -> List["ray.ObjectRef[R]"]:
+        "Call all the actors in the mesh with the same arguments."
+        pass
+
+    def choose(self, *args: P.args, **kwargs: P.kwargs) -> "ray.ObjectRef[R]":
+        "Call one of the actors in the mesh."
+        pass
+        
+    def shard(self, *args: P.args, **kwargs: P.kwargs) -> R:
+        "Call all the actors in the mesh, shard the argument(s) among them."
+        pass
+    
+class ActorMethodProxy:
+
+    def __init__(self, actor_mesh: "ActorMesh", actor_method: str) -> None:
+        self.actor_mesh = actor_mesh
+        self.actor_method = actor_method
+
+    def all(self, *args, **kwargs):
+        results = []
+        for actor in self.actor_mesh.actors:
+            method = getattr(actor, self.actor_method)
+            results.append(method.remote(*args, **kwargs))
+        return results
+    
+    def choose(self, *args, **kwargs):
+        # Choosing randomly right now, this is bad and needs to be improved
+        actor = random.choice(self.actor_mesh.actors)
```

Comment:
@jjyao @stephanie-wang @kevin85421 Do you know of a good way to do a better job with the load balancing here with existing Ray Core APIs?

Reply:
I don't think we currently expose number of pending tasks per actor but it wouldn't be very hard to add.

Seems also useful to have a version of `choose` that lets you pass in a specific index? It'd be nice to be able to this:

```python
mesh = ...
ref = mesh[1, 0].foo.remote(...)
```

Eventually could also use it for sub-meshes:
```python
2d_mesh = ...
sub_mesh = mesh[1, :]
```

Reply:
Also, the API that lets Ray choose the least-loaded could be reused for Ray Data actor pools. (cc @jjyao)

Reply:
That, or we promote the pool itself to a core API and implement it in cpp directly without adding a new low-level API. This will likely be needed to satisfy data FT requirements (retry on any actor in the pool)